Dreamacro/clash
# [Feature] Tun

This is a **Draft** PR, any comments are welcome.

**Plan**:
- [x] Tun Configuration
- [x] TCP support
- [x] UDP support
- [x] macOS support (5266519)
- [x] "Not support" message for windows and other OS
- [x] HTTP API support (GET, PATCH)]
- [x] DNS Service on Tun network
- [ ] Or support some other OS with [water](https://github.com/songgao/water)

**TODO**
- [ ] Rewrite UDP relay (working in #441）
- [ ] UDP Full-Cone NAT （Proof of concept works)  
 
**Unplan** for this PR
- changing routing table automatically
- fwmark

**Notice**: 

- For now, you need root privilege or CAP_NET_ADMIN to create tun NIC
- I use [google/netstack ](https://github.com/google/netstack) to handle TCP connection. The netstack **add 1.1MB extra size** to the clash binary.


Configure sample:
```yaml
tun:
  enable: true
  device-url: dev://utun # macOS
# device-url: dev://clash0 # Linux
# device-url: fd://5 # Linux
  dns-listen: 0.0.0.0:53  # additional dns server listen on TUN
```

**Test**（only a proof of concept of this PR):

1. Add tun config to `config.yaml`
2. Start clash
3. Bring tun interface up and add some route to routing table. make sure the target will not go through DIRECT in clash, or there will be a loop.
  e.g.`sudo ifconfig utun up; sudo route add 1.1.1.1 dev clash0;` (GNU/Linux)
  e.g. `sudo route add 1.1.1.1 -link utun2` (macOS)
4. Test it.
  e.g. `curl 1.1.1.1`
  e.g. `dig fb.com @1.1.1.1` （UDP)

**Usage**

Here are some pre-build [binaries](https://github.com/comzyh/clash/releases) of this PR in case you need it.

1. Modify configure file as described in **Test** section
2. Start clash, ensure tun interface is successfully attached. There should be log like this
  >INFO[0000] Tun adapter have interface name: utun 
3. Bring the Tun interface up `sudo ifconfig utun up`

If you are knowledgeable about Linux networking, you can add a route to the Tun interface and play with it. If not, you can try Fake-IP then.

4. If you wish to use **fake-ip** please add the route to clash device like `sudo route add -net 198.18.0.0/16 dev utun`. Please make sure Fake-IP is enabled in DNS settings, and you have already changed your DNS server to clash.

**Advanced Usage (Real IP)**
If you wish to use real-ip, Here is a [script](https://gist.github.com/Kr328/a81444f46e13c540a9b4c932ad1d482c)  write by @Kr328 that demonstrate that how to use **iptables**, **ip rule** and **fwmark** to redirect traffic to Tun interface and avoid loop.

And [Here ](https://comzyh.gitbook.io/clash/real-ip-tun-example)is a detailed action guide I write in Chinese.

**Tun DNS**

Clash For Android project needs the ability that clash creates a DNS server listens on specific IP:port on Tun TCP stack. So I add the DNS server on Tun to create a DNS server on Tun or hijack DNS request to some specific IP address.

This feature is disabled by default. If you wish to use this feature, please add `dns-listen` field under the *tun* config. For example, the following config will create a DNS server listen on `198.18.0.1:53`

```
tun:
  enable: true
  dns-listen: 198.18.0.1:53
  device-url: "dev://utun"
```
If you want to hijack any DNS request to port 53, you can use `0.0.0.0:53` instead. Please note that **this DNS server will only affect traffic that forward to clash Tun [NIC](https://en.wikipedia.org/wiki/Network_interface_controller)**
